ntp-udp/src/raw_socket.rs

```diff
@@ -27,62 +27,101 @@ pub(crate) fn cerr(t: libc::c_int) -> std::io::Result<libc::c_int> {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+#[repr(i32)]
+#[allow(clippy::enum_variant_names)]
+pub(crate) enum TimestampMethod {
+    #[allow(dead_code)]
+    SoTimestamp = libc::SO_TIMESTAMP,
+    #[cfg(target_os = "linux")]
+    #[allow(dead_code)]
+    SoTimestampns = libc::SO_TIMESTAMPNS,
+    #[cfg(target_os = "linux")]
+    SoTimestamping = libc::SO_TIMESTAMPING,
+}
+
 mod set_timestamping_options {
     use std::os::unix::prelude::AsRawFd;
 
-    use super::{cerr, TimestampingConfig};
+    use super::{cerr, TimestampMethod, TimestampingConfig};
+
+    fn configure_timestamping_socket(
+        udp_socket: &std::net::UdpSocket,
+        method: TimestampMethod,
+        options: u32,
+    ) -> std::io::Result<libc::c_int> {
+        // Documentation on the timestamping calls:
+        //
+        // - linux: https://www.kernel.org/doc/Documentation/networking/timestamping.txt
+        // - freebsd: https://man.freebsd.org/cgi/man.cgi?setsockopt
+        //
+        // SAFETY:
+        //
+        // - the socket is provided by (safe) rust, and will outlive the call
+        // - method is guaranteed to be a valid "name" argument
+        // - the options pointer outlives the call
+        // - the `option_len` corresponds with the options pointer
+        //
+        // Only some bits are valid to set in `options`, but setting invalid bits is perfectly safe
+        //
+        // > Setting other bit returns EINVAL and does not change the current state.
+        unsafe {
+            cerr(libc::setsockopt(
+                udp_socket.as_raw_fd(),
+                libc::SOL_SOCKET,
+                method as i32 as libc::c_int,
+                &options as *const _ as *const libc::c_void,
+                std::mem::size_of_val(&options) as libc::socklen_t,
+            ))
+        }
+    }
 
     pub(crate) fn set_timestamping_options(
         udp_socket: &std::net::UdpSocket,
+        method: TimestampMethod,
         timestamping: TimestampingConfig,
     ) -> std::io::Result<()> {
-        let fd = udp_socket.as_raw_fd();
-
-        let mut options = 0;
-
-        if timestamping.rx_software || timestamping.tx_software {
-            // enable software timestamping
-            options |= libc::SOF_TIMESTAMPING_SOFTWARE
-        }
+        let options = match method {
+            TimestampMethod::SoTimestamp => timestamping.rx_software as u32,
+            #[cfg(target_os = "linux")]
+            TimestampMethod::SoTimestampns => timestamping.rx_software as u32,
```

Comment:
These cases are somewhat unclear in what they do, maybe add a comment?

Reply:
Disagree with this being resolved, the type conversions still seem magical to me.